akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,436 @@
+          .to(Sink.ignore),
+          Promise[Done]().future) // never completed, not enabled
+    }
+
+    // An inbound connection will only use one of the control, ordinary or large streams, but we have to
+    // attach it to all and select via Partition and the streamId in the frame header. Conceptually it
+    // would have been better to send the streamId as one single first byte for a new connection and
+    // decide where to attach it based on that byte. Then the streamId wouldn't have to be sent in each
+    // frame. That was not chosen because it is more complicated to implement and might have more runtime
+    // overhead.
+    inboundStream =
+      OptionVal.Some(Sink.fromGraph(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+        val partition = b.add(Partition[EnvelopeBuffer](3, env ⇒ {
+          env.byteBuffer.get(EnvelopeBuffer.StreamIdOffset).toInt match {
+            case OrdinaryStreamId ⇒ 1
+            case ControlStreamId  ⇒ 0
+            case LargeStreamId    ⇒ 2
+            case other            ⇒ throw new IllegalArgumentException(s"Unexpected streamId [$other]")
+          }
+        }))
+        partition.out(0) ~> controlStream
+        partition.out(1) ~> ordinaryMessagesStream
+        partition.out(2) ~> largeMessagesStream
+        SinkShape(partition.in)
+      }))
+
+    val maxFrameSize =
+      if (largeMessageChannelEnabled) settings.Advanced.MaximumLargeFrameSize
+      else settings.Advanced.MaximumFrameSize
+
+    // TODO `Flow.alsoTo` is currently using eagerCancel = false, which is not what I want in the
+    // inboundConnectionFlow. If that is changed (see issue #24291) this custom impl can be replaced.
+    def alsoToEagerCancel[M](that: Graph[SinkShape[EnvelopeBuffer], M]): Graph[FlowShape[EnvelopeBuffer, EnvelopeBuffer], M] =
+      GraphDSL.create(that) { implicit b ⇒ r ⇒
+        import GraphDSL.Implicits._
+        val bcast = b.add(Broadcast[EnvelopeBuffer](2, eagerCancel = true))
+        bcast.out(1) ~> r
+        FlowShape(bcast.in, bcast.out(0))
+      }
+
+    // If something in the inboundConnectionFlow fails, e.g. framing, the connection will be teared down,
+    // but other parts of the inbound streams don't have to restarted.
+    def inboundConnectionFlow(): Flow[ByteString, ByteString, NotUsed] = {
+      // must create new Flow for each connection because of the FlightRecorder that can't be shared
+      val afr = createFlightRecorderEventSink()
+      Flow[ByteString]
+        .via(inboundKillSwitch.flow)
+        .via(Framing.lengthField(fieldLength = 4, fieldOffset = EnvelopeBuffer.FrameLengthOffset,
+          maxFrameSize, byteOrder = ByteOrder.LITTLE_ENDIAN))
+        .map { frame ⇒
+          val buffer = ByteBuffer.wrap(frame.toArray)
+          buffer.order(ByteOrder.LITTLE_ENDIAN)
+          afr.hiFreq(TcpInbound_Received, buffer.limit)
+          new EnvelopeBuffer(buffer)
```

Comment:
We should at some point try to remove a bit of copying buffers around. The whole concept of `EnvelopeBuffer` doesn't make sense on top of akka-stream TCP where we get the data delivered as a ByteString. So, now we copy this buffer into a `ByteBuffer` and then later copy the payload data back. Maybe we can change the envelope reader to not depend on `ByteBuffer` directly but use an abstraction to access bytes, so we can pass in the received ByteBuffer directly?

On the sending side we'll need one extra copy anyway to get the data into the ByteString to send.

Reply:
created https://github.com/akka/akka/issues/24556